geo/src/algorithm/lines_iter.rs

```diff
@@ -0,0 +1,328 @@
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        LineStringIter::new(self)
+    }
+}
+
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct LineStringIter<'a, T: CoordNum>(slice::Windows<'a, Coordinate<T>>);
+
+impl<'a, T: CoordNum> LineStringIter<'a, T> {
+    fn new(line_string: &'a LineString<T>) -> Self {
+        Self(line_string.0.windows(2))
+    }
+}
+
+impl<'a, T: CoordNum> Iterator for LineStringIter<'a, T> {
+    type Item = Line<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Can't use LineString::lines() because it returns an `impl Trait`
+        // and there is no way to name that type in `LinesIter::Iter` until [RFC 2071] is stabilized.
+        //
+        // [RFC 2071]: https://rust-lang.github.io/rfcs/2071-impl-trait-existential-types.html
+        self.0.next().map(|w| {
+            // slice::windows(2) is guaranteed to yield a slice with exactly 2 elements
+            unsafe { Line::new(*w.get_unchecked(0), *w.get_unchecked(1)) }
+        })
+    }
+}
+
+// ┌───────────────────────────────────────┐
+// │ Implementation for MultiLineString    │
+// └───────────────────────────────────────┘
+
+type MultiLineStringIter<'a, T> =
+    iter::Flatten<MapLinesIter<'a, slice::Iter<'a, LineString<T>>, LineString<T>>>;
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for MultiLineString<T> {
+    type Scalar = T;
+    type Iter = MultiLineStringIter<'a, Self::Scalar>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        MapLinesIter(self.0.iter()).flatten()
+    }
+}
+
+// ┌───────────────────────────────┐
+// │ Implementation for Polygon    │
+// └───────────────────────────────┘
+
+type PolygonIter<'a, T> = iter::Chain<
+    LineStringIter<'a, T>,
+    iter::Flatten<MapLinesIter<'a, slice::Iter<'a, LineString<T>>, LineString<T>>>,
+>;
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for Polygon<T> {
+    type Scalar = T;
+    type Iter = PolygonIter<'a, Self::Scalar>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        self.exterior()
+            .lines_iter()
+            .chain(MapLinesIter(self.interiors().iter()).flatten())
+    }
+}
+
+// ┌────────────────────────────────────┐
+// │ Implementation for MultiPolygon    │
+// └────────────────────────────────────┘
+
+type MultiPolygonIter<'a, T> =
+    iter::Flatten<MapLinesIter<'a, slice::Iter<'a, Polygon<T>>, Polygon<T>>>;
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for MultiPolygon<T> {
+    type Scalar = T;
+    type Iter = MultiPolygonIter<'a, Self::Scalar>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        MapLinesIter(self.0.iter()).flatten()
+    }
+}
+
+// ┌────────────────────────────┐
+// │ Implementation for Rect    │
+// └────────────────────────────┘
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for Rect<T> {
+    type Scalar = T;
+    type Iter = <[Line<Self::Scalar>; 4] as IntoIterator>::IntoIter;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        // Explicitly iterate by value so this works for pre-2021 rust editions.
```

Comment:
We might switch to 2021 in #741.

Reply:
Should I add a TODO in the comment referencing the PR? I can remove it once 741 is merged.

Reply:
Added TODO.

Will keep an eye on #741 and #758 and update this once those are merged.

Reply:
To double check, we'll still need #741 to be merged, right? Because #758 doesn't actually bump MSRV, so our clients could still be on pre-2021 rustc.

Reply:
FYI #741 is building now. I expect it to be merged imminently. 

Sorry for the rigamarole as all the trains come into the station at once...

Reply:
No worries. I'm happy to have converted a TODO to a to-done ;)